### ISSUE TITLE
Tracking issue for displaced tracks and electron reconstruction efficiency

### ISSUE THREAD

**Issue by gaswk** [key4hep/CLDConfig#25]
After investigating the tracking issue, adjusting the parameters `MaxCellAngle` and `MaxCellAngleRZ `to 0.5 (as opposed to 0.1), in the `displaced step`, has significantly enhanced the reconstruction of displaced tracks (tested with displaced single particle). These [conditions](https://github.com/iLCSoft/ConformalTracking/blob/master/src/ConformalTracking.cc#L1686) were frequently not met, leading to track not being reconstructed.
The choice of the 0.5 value was somewhat arbitrary, and further optimisation is required.

However, despite this adjustment, an issue persists in properly reconstructing the Heavy Neutral Lepton (HNL) vertex with electron tracks. Interestingly, when considering all tracks in the event, the HNL vertex reconstruction aligns better with the expected distribution. What factors could be contributing to the reconstruction efficiency of electron?

1- Reconstructed vertex displacement (normalised to 1)
![Screenshot from 2024-03-07 13-44-19](https://github.com/key4hep/CLDConfig/assets/116810451/2dcaa518-9644-49df-a28c-eb60beff2130)

2- Comparison of Generated and reconstructed vertex displacement for `MaxCellAngle`, `MaxCellAngleRZ` = 0.5, left reconstructed with electrons tracks and right with all tracks in the event
![Screenshot from 2024-03-07 13-46-34](https://github.com/key4hep/CLDConfig/assets/116810451/98bf418f-ffb9-4b52-810e-c8cfb0ee19c4)  ![Screenshot from 2024-03-07 13-47-09](https://github.com/key4hep/CLDConfig/assets/116810451/ecb19fef-136a-4d55-846f-86922a7bd942)

3- Comparison of Generated and reconstructed vertex displacement `MaxCellAngle`, `MaxCellAngleRZ` = 0.1, left reconstructed with electrons tracks and right with all tracks in the event
![Screenshot from 2024-03-07 13-48-38](https://github.com/key4hep/CLDConfig/assets/116810451/4467ba6e-4649-433f-9987-5f88d52704cf)  ![Screenshot from 2024-03-07 13-48-48](https://github.com/key4hep/CLDConfig/assets/116810451/0750afd9-4804-40c6-8c6e-30225b18e1ce)

**Comment by Zehvogel** [comment]
Hi Galle,

very interesting results!
Can you also check how the change of this criterion affects the track finding efficiency and the fake rate?
It is a bit hard to judge from your plots if the distribution is different because the tracks are missing or because the tracks are reconstructed incorrectly.
In the longer run we should probably collect a set of tracking performance plot scripts that we can put into a CI workflow...

Regarding the electrons: there are also some issues reconstructing prompt electrons that I noticed recently
![image](https://github.com/key4hep/CLDConfig/assets/12029880/1282b841-4b6d-47f4-971b-dae7fc975827)
I hoped that this can be fixed by relaxing another cut somewhere, but I had no time to investigate yet.

**Comment by gaswk** [comment]
I gave you access to the aida files, with both values of  `MaxCellAngle`. I think tracks were not correctly reconstructed. Both files are for 500mm displayed muons

/eos/user/g/gasadows/Output/TrackingPerformance/testREC_500mm_MCA01_aida.root

/eos/user/g/gasadows/Output/TrackingPerformance/testREC_500mm_MCA05_aida.root

**Comment by Zehvogel** [comment]
Yes, as far as I can see from the files (`MyClicEfficiencyCalculator/simplifiedEfficiencyTree/m_reconstructed`) the issue seems to be that the tracks are not reconstructed at all.

In the 0.1 case its only 6/87 tracks and in the 0.5 case it is 37/87

I am a bit surprised by the uneven number of 87 but I think this includes being in the detector acceptance and maybe one of the particles goes down the beamline... 

I would still like to check how this change influences the fake rate of tracking in more busy events before making this the new default but that will still take me a while.

**Comment by gaswk** [comment]
While investigating Pandora particles, mainly photons instead of electrons, tracks are not consistently associated with clusters. A comparison of track positions at calo and cluster positions reveals a mismatch.
![Screenshot from 2024-07-29 09-49-47](https://github.com/user-attachments/assets/3fb7539e-ff69-42ba-90b5-8a6d75b3bef9)
![Screenshot from 2024-07-29 09-50-11](https://github.com/user-attachments/assets/122e6d38-7139-4786-ac7b-01bff8d24210)

**Comment by andresailer** [comment]
Hi @gaswk 

Can you plot the 3D distance between tracks and clusters? Maybe even split between tracks and clusters that were matched and unmatched clusters and tracks?
The track position is at the calo front face surface, and the cluster in the calorimeter, so some distance is expected.

**Comment by gaswk** [comment]
Here is DeltaR and DeltaD for tracks and clusters position
![Screenshot from 2024-08-06 14-03-56](https://github.com/user-attachments/assets/92b5f201-759b-49e2-91eb-2316e04fb7ee)

**Comment by andresailer** [comment]
Thanks!

What is DeltaD? 
And DeltaR is the difference in radius between trackState and cluster, or cluster and trackstate?

**Comment by gaswk** [comment]
D = sqrt(X^2 + Y^2 + Z^2)
Both are differences between cluster and trackstate (Delta = cluster - tracktate)

**Comment by andresailer** [comment]
It is `DeltaD = sqrt(Sum_i (X_c^i-X_t^i)^2))` ? How can it be negative?

**Comment by gaswk** [comment]
`DeltaD = D_cluster - D_trackstate`

**Comment by andresailer** [comment]
That is not the 3D distance though. So if you could plot the DeltaD as I defined above?

**Comment by gaswk** [comment]
Here is the Delta3D distribution for matched and not matched electrons
![Screenshot from 2024-08-13 18-51-07](https://github.com/user-attachments/assets/2b9e2ad2-c0f5-4af6-ab1f-2136bfc580ff)

**Comment by andresailer** [comment]
Thanks for the plot @gaswk ! There is not really any difference between the distribution, so the issue is probably elsewhere.

**Comment by gaswk** [comment]
I have conducted a more detailed study on the `MaxCellAngle` values and their impact. Below are the results:

Efficiencies (as a function of `Gen vertex Lxyz`, `pT`, and `theta`) for 50k 50 GeV HNL events:

![Image](https://github.com/user-attachments/assets/302a6c71-9b63-43b0-bd38-11cdea102eb5)
![Image](https://github.com/user-attachments/assets/ffee772e-a74e-4661-ac11-578922efeae3)
![Image](https://github.com/user-attachments/assets/b372c7da-641e-4a20-be86-2564e74ccd7a)

Fake rates (as a function of the `distance at the first hit`, `pT`, and `theta`) for 50k Zbb events:

![Image](https://github.com/user-attachments/assets/02258c6e-d166-4bd7-8ed5-8a5520109f44)
![Image](https://github.com/user-attachments/assets/ddc17953-5d54-49b6-9e6b-203a173d1842)
![Image](https://github.com/user-attachments/assets/6341db9c-c3e8-4aca-812c-aaa857684380)

Reconstruction time comparison for `MaxCellAngle=0.1` and `MaxCellAngle=0.5` (1k Zbb events):
`MaxCellAngle=0.1`
real	        6m25.744s
user	6m7.307s
sys	       0m8.912s

`MaxCellAngle=0.5`
real	        8m45.309s
user	8m29.495s
sys	        0m7.956s

**Comment by Zehvogel** [comment]
Do I read your plots correctly that we have a 10% fake rate for tracks at the moment even without the change? That sounds concerning. 😰 

In the detector note we had ~1% (under slightly different conditions, but still...)

![Image](https://github.com/user-attachments/assets/7ac03e77-88c1-4834-bc1b-39643fac5b5a)

**Comment by gaswk** [comment]
Apologies for not clarifying earlier. Here are the definitions I used:

- Fake Rate: fake tracks (weight in mctruthrecolink_trk < 0.99)) / all reconstructed tracks
- Efficiency: MC-matched electrons / all MC electrons

**Comment by andresailer** [comment]
> two tracks link to the same MC

Those are not fake tracks, those might be split tracks.
